File: src/Arqanore/Graphics/Texture.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Text;
using Arqan;
using Arqanore.Utils;

namespace Arqanore.Graphics
{
    public class Texture
    {
        public uint Id { get; private set; }
        public int Width { get; private set; }
        public int Height { get; private set; }
        public Bitmap Bitmap { get; private set; }
        public IntPtr Pixels { get; private set; }
        public PixelFormat PixelFormat { get; private set; }

        public Texture(Image img)
        {
            Load(new Bitmap(img));
        }
        public Texture(Bitmap bmp)
        {
            Load(bmp);
        }
        public Texture(string filename)
        {
            // Check the file
            if (!filename.EndsWith(".arqtex"))
            {
                throw new ArqanoreException("Not a valid Arqanore texture");
            }
            if(!File.Exists(filename))
            {
                throw new ArqanoreException($"File {filename} not found");
            }

            // Parse the data
            Parse(File.ReadAllBytes(filename));
        }
        public Texture(byte[] data)
        {
            Parse(data);
        }

        private void Parse(byte[] data)
        {
            var parser = new ByteParser(data);
            var img = parser.GetImage(data.Length);
            var bmp = new Bitmap(img);

            Load(bmp);
        }
        private void Load(Bitmap bmp)
        {
            var ids = new uint[1];
            var data = bmp.LockBits(new Rectangle(0, 0, bmp.Width, bmp.Height), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);

            // Set properties
            Width = bmp.Width;
            Height = bmp.Height;
            Bitmap = bmp;
            Pixels = data.Scan0;
            PixelFormat = data.PixelFormat;

            // Generate 2D texture
            GL.glGenTextures(1, ids);
            GL.glBindTexture(GL.GL_TEXTURE_2D, ids[0]);
            GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_WRAP_S, GL.GL_REPEAT);
            GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_WRAP_T, GL.GL_REPEAT);
            GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_NEAREST);
            GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_NEAREST);
            GL.glTexImage2D(GL.GL_TEXTURE_2D, 0, GL.GL_RGBA, bmp.Width, bmp.Height, 0, GL.GL_BGRA, GL.GL_UNSIGNED_BYTE, Pixels);

            // Cleanup
            bmp.UnlockBits(data);
            GL.glBindTexture(GL.GL_TEXTURE_2D, 0);

            // Finish
            Id = ids[0];
        }

        public Color GetPixel(int x, int y)
        {
            return new Color(Bitmap.GetPixel(x, y));
        }
    }
}

File: src/arqanore/texture.cpp
#include "arqanore/texture.h"
#include "stb/stb_image.h"
#include "glad/gl.h"
#include "arqanore/exceptions.h"

arqanore::Texture::Texture() {
    id = 0;
    width = 0;
    height = 0;
};

arqanore::Texture::Texture(std::string filename) {
    int channels, width, height;
    unsigned char *data = stbi_load(filename.c_str(), &width, &height, &channels, STBI_rgb_alpha);

    if (!data) {
        std::string reason = stbi_failure_reason();
        std::string error = "Failed to load texture '" + (std::string) filename + "'. Reason is '" + reason + "'";

        stbi_image_free(data);
        throw ArqanoreException(error);
    }

    glGenTextures(1, &this->id);
    glBindTexture(GL_TEXTURE_2D, this->id);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, data);

    stbi_image_free(data);
    this->width = width;
    this->height = height;
}
